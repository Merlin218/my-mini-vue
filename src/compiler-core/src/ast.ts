import { CREATE_ELEMENT_VNODE } from "./runtimeHelpers"

export const enum NodeTypes {
  INTERPOLATION = 'interpolation',
  SIMPLE_EXPRESSION = 'simply_expression',
  ELEMENT = 'element',
  TEXT = 'text',
  ROOT = 'root',
  COMPOUND_EXPRESSION = 'compound_expression'
}

export function createVNodeCall(ctx,tag, props, children) {
  ctx.helper(CREATE_ELEMENT_VNODE)
  
  return {
    type: NodeTypes.ELEMENT,
    tag, props, children
  }
}
