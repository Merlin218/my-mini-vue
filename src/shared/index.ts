export function isObject(target) {
  return typeof target === "object" && target !== null;
}

export function isReactive(target) {
  return !!(target && target.__isReactive);
}
