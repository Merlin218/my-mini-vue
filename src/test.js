import { toDisplayString as _toDisplayString, createElementVNode as _createElementVNode } from 'Vue';
return function render(_ctx, _cache){return _createElementVNode('div', null, 'hi,' + _toDisplayString(_ctx.message))}
