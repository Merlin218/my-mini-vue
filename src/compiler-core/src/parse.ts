import { NodeTypes } from "./ast";

const enum TagType {
  Start = 'tag_start',
  End = 'tag_end'
}

export function baseParse(context: string) {
  const content = createParseContext(context)
  return createRoot(parseChildren(content))
}

function parseChildren(context) {
  const nodes = [];
  let node: any;
  const s = context.source;
  if (s.startsWith("{{")) {
    node = parseInterpolation(context)
  } else if (s[0] === '<') {
    if (/[a-z]/i.test(s[1])) {
      node = parseElement(context);
    }
  }

  if (!node) {
    node = parseText(context)
  }

  nodes.push(node)

  return nodes;
}

function parseText(context) {
  const content = parseTextData(context,context.source.length);
  return {
    type: NodeTypes.TEXT,
    content,
  }
}

function parseTextData(context: any,length:number) {
  const content = context.source.slice(0, length);
  advanceBy(context, length);
  return content;
}

function parseElement(context) {
  const element = parseTag(context, TagType.Start);
  parseTag(context, TagType.End)
  return element;
}

function parseTag(context, type) {
  const match: any = /^<\/?([a-z]*)>/i.exec(context.source);

  const tag = match[1];
  advanceBy(context, match[0].length);

  if (type === TagType.End) return;
  return {
    type: NodeTypes.ELEMENT,
    tag
  }
}

function parseInterpolation(context) {

  const openDelimiter = '{{';
  const closeDelimiter = '}}';

  // 拿到结尾的索引
  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)
  // 前进，去除{{
  advanceBy(context, openDelimiter.length);
  // 目标内容的长度
  const rawContentLength = closeIndex - openDelimiter.length
  // 获取的内容
  const rawContent = parseTextData(context,rawContentLength)
  const content = rawContent.trim()

  // 前进，去除}}
  advanceBy(context, closeDelimiter.length);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    }
  }
}


function advanceBy(context, length) {
  context.source = context.source.slice(length);
}

function createRoot(children) {
  return {
    children,
  }
}

function createParseContext(content: string) {
  return {
    source: content
  }
}
