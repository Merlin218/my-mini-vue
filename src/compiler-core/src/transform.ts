import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";

export function transform(root, options = {}) {

  const context = createTransformContext(root, options);

  // 1. 遍历 - 深度优先搜索
  traverseNode(root, context)
  // 2. 修改 text content

  createRootCodegen(root)
  root.helpers = [...context.helpers.keys()]
}

function createRootCodegen(root) {
  const child = root.children[0];
  if(child.type === NodeTypes.ELEMENT){
    root.codegenNode = child.codegenNode
  }else{
    root.codegenNode = root.children[0];
  }
}


function traverseNode(node, context) {
  const { nodeTransforms } = context
  let exitFns:any = [];
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    // 调用中间层之后，可能会返回一个函数，那么储存起来
    // 在逻辑处理完成后再进行调用
    // 例如：实现后的调用逻辑为[1,2,3,3',2',1']
    const onExit =  transform(node, context);
    if(onExit) exitFns.push(onExit);
  }
  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING)
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChild(node, context);
      break;
    default:
      break;
  }
  // 流程结束之后，会取出所有中间件返回的函数，倒序执行
  let i = exitFns.length
  while(i--){
    exitFns[i]()
  }
}

function traverseChild(node: any, context: any) {
  const children = node.children;
  if (children) {
    for (let i = 0; i < children.length; i++) {
      const node = children[i];
      traverseNode(node, context);
    }
  }
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      this.helpers.set(key, 1);
    }
  }
  return context;
}
