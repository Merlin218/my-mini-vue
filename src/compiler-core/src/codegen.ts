import { isString } from "../../shared";
import { NodeTypes } from "./ast";
import { CREATE_ELEMENT_VNODE, helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers";

export function generate(ast) {
  const context = createCodegenContext()
  const { push } = context

  genFunctionPreamble(ast, context);

  const functionName = "render"
  const args = ["_ctx", "_cache"]
  push(`return `)

  const signature = args.join(', ');


  push(`function ${functionName}(${signature}){`);
  push(`return `)
  genNode(ast.codegenNode, context);
  push('}')

  return {
    code: context.code
  }
}
function genFunctionPreamble(ast: any, context: any) {
  const { push } = context
  const VueBinging = 'Vue';
  const aliasHelpers = (i) => `${helperMapName[i]}: _${helperMapName[i]}`;
  if (ast.helpers.length > 0) {
    push(`const { ${ast.helpers.map(aliasHelpers).join(", ")} } = ${VueBinging}`);
    push('\n');
  }
}

function genNode(node: any, ctx: any) {
  if (!node) return;
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, ctx);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, ctx);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, ctx);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, ctx);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, ctx);
      break;
    default:
      break;
  }
}

function genCompoundExpression(node, ctx) {
  const { children } = node;
  const { push } = ctx
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (isString(child)) {
      push(child)
    } else {
      genNode(child, ctx)
    }

  }
}

function genElement(node, ctx) {
  const { push, helper } = ctx;
  const { tag, children, props } = node
  const child = children[0];
  push(`${helper(CREATE_ELEMENT_VNODE)}(`)
  genNodeList(genNullable([tag, props, children]), ctx)
  // genNode(children,ctx)
  push(')')
}

function genNodeList(nodes, ctx) {
  const { push } = ctx
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      push(node)
    } else {
      genNode(node, ctx)
    }
    if (i < nodes.length - 1) push(', ')
  }
}

function genNullable(args) {
  return args.map((arg) => arg || 'null')
}

function genText(node: any, ctx: any) {
  const { push } = ctx;
  push(`'${node.content}'`);
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapName[key]}`
    }
  }
  return context;
}


function genInterpolation(node: any, ctx: any) {
  const { push, helper } = ctx;
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, ctx)
  push(')')
}

function genExpression(node: any, ctx: any) {
  const { push } = ctx
  push(`${node.content}`)
}
