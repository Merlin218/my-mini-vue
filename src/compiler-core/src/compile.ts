import { generate } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformElement } from "./transform/transformElement";
import { transformExpression } from "./transform/transformExpression";
import { transformText } from "./transform/transformText";

export function baseComplie(template){
  const ast: any = baseParse(template);

  // 涉及一个中间层的调用顺序的问题
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText]
  })
  return generate(ast);
}
