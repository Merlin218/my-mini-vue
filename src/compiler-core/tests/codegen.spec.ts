import { generate } from "../src/codegen";
import { baseParse } from "../src/parse"
import { transform } from "../src/transform";
import { transformExpression } from "../src/transform/transformExpression";
import { transformElement } from "../src/transform/transformElement";
import { transformText } from "../src/transform/transformText";

describe('codepen', () => {
  it('string', () => {
    const ast = baseParse('hi');

    transform(ast)

    const { code } = generate(ast);

    // 快照
    expect(code).toMatchSnapshot();
  })

  it('interpolation', () => {
    const ast = baseParse('{{message}}');

    transform(ast, {
      nodeTransforms: [transformExpression]
    })

    const { code } = generate(ast);

    // 快照
    expect(code).toMatchSnapshot();
  })

  it('element', () => {
    const ast: any = baseParse('<div>hi,{{message}}</div>');

    // 涉及一个中间层的调用顺序的问题
    transform(ast, {
      nodeTransforms: [transformExpression, transformElement, transformText]
    })
    const { code } = generate(ast);

    // 快照
    expect(code).toMatchSnapshot();
  })
})
