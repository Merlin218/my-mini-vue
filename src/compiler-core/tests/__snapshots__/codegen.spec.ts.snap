// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codepen element 1`] = `
"import { toDisplayString: _toDisplayString, createElementVNode: _createElementVNode } from Vue
return function render(_ctx, _cache){return _createElementVNode('div', null, 'hi,' + _toDisplayString(_ctx.message))}"
`;

exports[`codepen interpolation 1`] = `
"import { toDisplayString: _toDisplayString } from Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"
`;

exports[`codepen string 1`] = `"return function render(_ctx, _cache){return 'hi'}"`;
